# Problem Statement

## **Business Context**

"Visit with Us," a leading travel company, is revolutionizing the tourism industry by leveraging data-driven strategies to optimize operations and customer engagement. While introducing a new package offering, such as the Wellness Tourism Package, the company faces challenges in targeting the right customers efficiently. The manual approach to identifying potential customers is inconsistent, time-consuming, and prone to errors, leading to missed opportunities and suboptimal campaign performance.

To address these issues, the company aims to implement a scalable and automated system that integrates customer data, predicts potential buyers, and enhances decision-making for marketing strategies. By utilizing an MLOps pipeline, the company seeks to achieve seamless integration of data preprocessing, model development, deployment, and CI/CD practices for continuous improvement. This system will ensure efficient targeting of customers, timely updates to the predictive model, and adaptation to evolving customer behaviors, ultimately driving growth and customer satisfaction.


## **Objective**

As an MLOps Engineer at "Visit with Us," your responsibility is to design and deploy an MLOps pipeline on GitHub to automate the end-to-end workflow for predicting customer purchases. The primary objective is to build a model that predicts whether a customer will purchase the newly introduced Wellness Tourism Package before contacting them. The pipeline will include data cleaning, preprocessing, transformation, model building, training, evaluation, and deployment, ensuring consistent performance and scalability. By leveraging GitHub Actions for CI/CD integration, the system will enable automated updates, streamline model deployment, and improve operational efficiency. This robust predictive solution will empower policymakers to make data-driven decisions, enhance marketing strategies, and effectively target potential customers, thereby driving customer acquisition and business growth.

## **Data Description**

The dataset contains customer and interaction data that serve as key attributes for predicting the likelihood of purchasing the Wellness Tourism Package. The detailed attributes are:

**Customer Details**
- **CustomerID:** Unique identifier for each customer.
- **ProdTaken:** Target variable indicating whether the customer has purchased a package (0: No, 1: Yes).
- **Age:** Age of the customer.
- **TypeofContact:** The method by which the customer was contacted (Company Invited or Self Inquiry).
- **CityTier:** The city category based on development, population, and living standards (Tier 1 > Tier 2 > Tier 3).
- **Occupation:** Customer's occupation (e.g., Salaried, Freelancer).
- **Gender:** Gender of the customer (Male, Female).
- **NumberOfPersonVisiting:** Total number of people accompanying the customer on the trip.
- **PreferredPropertyStar:** Preferred hotel rating by the customer.
- **MaritalStatus:** Marital status of the customer (Single, Married, Divorced).
- **NumberOfTrips:** Average number of trips the customer takes annually.
- **Passport:** Whether the customer holds a valid passport (0: No, 1: Yes).
- **OwnCar:** Whether the customer owns a car (0: No, 1: Yes).
- **NumberOfChildrenVisiting:** Number of children below age 5 accompanying the customer.
- **Designation:** Customer's designation in their current organization.
- **MonthlyIncome:** Gross monthly income of the customer.

**Customer Interaction Data**
- **PitchSatisfactionScore:** Score indicating the customer's satisfaction with the sales pitch.
- **ProductPitched:** The type of product pitched to the customer.
- **NumberOfFollowups:** Total number of follow-ups by the salesperson after the sales pitch.-
- **DurationOfPitch:** Duration of the sales pitch delivered to the customer.


# Model Building

In [1]:
# Create a master folder to keep all files created when executing the below code cells
import os
os.makedirs("tourism_project", exist_ok=True)

In [2]:
# Create a folder for storing the model building files
os.makedirs("tourism_project/model_building", exist_ok=True)

## Data Registration

In [3]:
os.makedirs("tourism_project/data", exist_ok=True)

In [4]:
import pandas as pd
df = pd.read_csv("tourism.csv")
df.head()

,Unnamed: 0,CustomerID,ProdTaken,Age,TypeofContact,CityTier,DurationOfPitch,Occupation,Gender,NumberOfPersonVisiting,...,ProductPitched,PreferredPropertyStar,MaritalStatus,NumberOfTrips,Passport,PitchSatisfactionScore,OwnCar,NumberOfChildrenVisiting,Designation,MonthlyIncome
0,0,200000,1,41.0,Self Enquiry,3,6.0,Salaried,Female,3,...,Deluxe,3.0,Single,1.0,1,2,1,0.0,Manager,20993.0
1,1,200001,0,49.0,Company Invited,1,14.0,Salaried,Male,3,...,Deluxe,4.0,Divorced,2.0,0,3,1,2.0,Manager,20130.0
2,2,200002,1,37.0,Self Enquiry,1,8.0,Free Lancer,Male,3,...,Basic,3.0,Single,7.0,1,3,0,0.0,Executive,17090.0
3,3,200003,0,33.0,Company Invited,1,9.0,Salaried,Female,2,...,Basic,3.0,Divorced,2.0,1,5,1,1.0,Executive,17909.0
4,5,200005,0,32.0,Company Invited,1,8.0,Salaried,Male,3,...,Basic,3.0,Single,1.0,0,5,1,1.0,Executive,18068.0


Once the **data** folder created after executing the above cell, please upload the **tourism.csv** in to the folder

## Data Preparation

In [5]:
import pandas as pd
df = pd.read_csv("tourism.csv")
df.head()

,Unnamed: 0,CustomerID,ProdTaken,Age,TypeofContact,CityTier,DurationOfPitch,Occupation,Gender,NumberOfPersonVisiting,...,ProductPitched,PreferredPropertyStar,MaritalStatus,NumberOfTrips,Passport,PitchSatisfactionScore,OwnCar,NumberOfChildrenVisiting,Designation,MonthlyIncome
0,0,200000,1,41.0,Self Enquiry,3,6.0,Salaried,Female,3,...,Deluxe,3.0,Single,1.0,1,2,1,0.0,Manager,20993.0
1,1,200001,0,49.0,Company Invited,1,14.0,Salaried,Male,3,...,Deluxe,4.0,Divorced,2.0,0,3,1,2.0,Manager,20130.0
2,2,200002,1,37.0,Self Enquiry,1,8.0,Free Lancer,Male,3,...,Basic,3.0,Single,7.0,1,3,0,0.0,Executive,17090.0
3,3,200003,0,33.0,Company Invited,1,9.0,Salaried,Female,2,...,Basic,3.0,Divorced,2.0,1,5,1,1.0,Executive,17909.0
4,5,200005,0,32.0,Company Invited,1,8.0,Salaried,Male,3,...,Basic,3.0,Single,1.0,0,5,1,1.0,Executive,18068.0


In [6]:
# Check the shape of the dataset
print("Dataset Shape:", df.shape)

# Check column names
print("\nColumn Names:")
print(df.columns.tolist())

# Check basic information about the dataset
print("\nDataset Information:")
df.info()

Dataset Shape: (4128, 21)

Column Names:
['Unnamed: 0', 'CustomerID', 'ProdTaken', 'Age', 'TypeofContact', 'CityTier', 'DurationOfPitch', 'Occupation', 'Gender', 'NumberOfPersonVisiting', 'NumberOfFollowups', 'ProductPitched', 'PreferredPropertyStar', 'MaritalStatus', 'NumberOfTrips', 'Passport', 'PitchSatisfactionScore', 'OwnCar', 'NumberOfChildrenVisiting', 'Designation', 'MonthlyIncome']

Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4128 entries, 0 to 4127
Data columns (total 21 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Unnamed: 0                4128 non-null   int64  
 1   CustomerID                4128 non-null   int64  
 2   ProdTaken                 4128 non-null   int64  
 3   Age                       4128 non-null   float64
 4   TypeofContact             4128 non-null   object 
 5   CityTier                  4128 non-null   int64  
 6   DurationOfPitch           4128 n

### Initial Data Quality Check

The dataset contains 4,128 observations and 21 columns. The data types include numerical and categorical features. No missing values are present based on the dataset information.

We will further check for duplicate records, target class distribution, and the unique values of categorical variables before performing preprocessing.

In [7]:
# Check for missing values
print("Missing Values:")
print(df.isnull().sum())

# Check for duplicate rows
print("\nNumber of Duplicate Rows:", df.duplicated().sum())

# Check target variable distribution
print("\nTarget Variable Distribution:")
print(df["ProdTaken"].value_counts())

print("\nTarget Variable Percentage:")
print(df["ProdTaken"].value_counts(normalize=True) * 100)

# Check unique values of categorical columns
categorical_cols = df.select_dtypes(include="object").columns

print("\nUnique Values in Categorical Columns:")
for col in categorical_cols:
    print(f"\n{col}:")
    print(df[col].unique())

Missing Values:
Unnamed: 0                  0
CustomerID                  0
ProdTaken                   0
Age                         0
TypeofContact               0
CityTier                    0
DurationOfPitch             0
Occupation                  0
Gender                      0
NumberOfPersonVisiting      0
NumberOfFollowups           0
ProductPitched              0
PreferredPropertyStar       0
MaritalStatus               0
NumberOfTrips               0
Passport                    0
PitchSatisfactionScore      0
OwnCar                      0
NumberOfChildrenVisiting    0
Designation                 0
MonthlyIncome               0
dtype: int64

Number of Duplicate Rows: 0

Target Variable Distribution:
ProdTaken
0    3331
1     797
Name: count, dtype: int64

Target Variable Percentage:
ProdTaken
0    80.692829
1    19.307171
Name: proportion, dtype: float64

Unique Values in Categorical Columns:

TypeofContact:
['Self Enquiry' 'Company Invited']

Occupation:
['Salaried' 'Free La

### Data Quality Findings

The dataset contains no missing values and no duplicate rows.

The target variable `ProdTaken` is imbalanced, with 80.69% of customers not purchasing the package and 19.31% purchasing it. Therefore, class imbalance should be considered while training and evaluating the classification models.

Among the categorical variables, the `Gender` column contains an inconsistent category, `Fe Male`, which should be standardized with `Female` before model training.

The `Unnamed: 0` column appears to be an index column and is not expected to provide useful predictive information. `CustomerID` is an identifier and should also be evaluated before being used as a model feature.

In [8]:
# Check whether CustomerID is unique
print("Unique CustomerID values:", df["CustomerID"].nunique())
print("Total rows:", len(df))

# Check numerical feature statistics
print("\nNumerical Feature Summary:")
print(df.describe().T)

# Check value counts for Gender
print("\nGender Distribution:")
print(df["Gender"].value_counts())

Unique CustomerID values: 4128
Total rows: 4128

Numerical Feature Summary:
                           count           mean          std       min  \
Unnamed: 0                4128.0    2527.763808  1409.439133       0.0   
CustomerID                4128.0  202527.763808  1409.439133  200000.0   
ProdTaken                 4128.0       0.193072     0.394757       0.0   
Age                       4128.0      37.231831     9.174521      18.0   
CityTier                  4128.0       1.663275     0.920640       1.0   
DurationOfPitch           4128.0      15.584787     8.398142       5.0   
NumberOfPersonVisiting    4128.0       2.949370     0.718818       1.0   
NumberOfFollowups         4128.0       3.741521     1.006786       1.0   
PreferredPropertyStar     4128.0       3.578488     0.795031       3.0   
NumberOfTrips             4128.0       3.295300     1.856300       1.0   
Passport                  4128.0       0.295300     0.456233       0.0   
PitchSatisfactionScore    4128.0    

### Data Cleaning Decisions

The `CustomerID` column contains a unique value for every observation and is an identifier rather than a predictive feature. The `Unnamed: 0` column is also an index-like column. Both columns will therefore be removed before model training.

The `Gender` column contains the categories `Male`, `Female`, and `Fe Male`. The `Fe Male` category is an inconsistent representation of `Female`, so it will be standardized to `Female`.

The numerical summary also shows some potentially high values in features such as `DurationOfPitch` and `MonthlyIncome`. These will be examined carefully rather than removed automatically.

In [9]:
# Create a copy for data preparation
df_clean = df.copy()

# Remove identifier/index columns
df_clean.drop(columns=["Unnamed: 0", "CustomerID"], inplace=True)

# Standardize inconsistent Gender category
df_clean["Gender"] = df_clean["Gender"].replace({"Fe Male": "Female"})

# Check the cleaned structure
print("Shape after removing identifier columns:", df_clean.shape)

print("\nGender values after standardization:")
print(df_clean["Gender"].value_counts())

print("\nRemaining columns:")
print(df_clean.columns.tolist())

Shape after removing identifier columns: (4128, 19)

Gender values after standardization:
Gender
Male      2463
Female    1665
Name: count, dtype: int64

Remaining columns:
['ProdTaken', 'Age', 'TypeofContact', 'CityTier', 'DurationOfPitch', 'Occupation', 'Gender', 'NumberOfPersonVisiting', 'NumberOfFollowups', 'ProductPitched', 'PreferredPropertyStar', 'MaritalStatus', 'NumberOfTrips', 'Passport', 'PitchSatisfactionScore', 'OwnCar', 'NumberOfChildrenVisiting', 'Designation', 'MonthlyIncome']


### Exploratory Feature Analysis

After removing the identifier columns and standardizing the inconsistent gender category, the remaining features are divided into numerical and categorical variables.

Before applying transformations, the distributions of numerical variables and the frequency of categorical variables will be examined to identify potential skewness, unusual values, or categories that may require special handling.

In [10]:
# Separate target variable
target = "ProdTaken"

# Identify numerical and categorical columns
numerical_cols = df_clean.drop(columns=[target]).select_dtypes(
    include=["int64", "float64"]
).columns.tolist()

categorical_cols = df_clean.drop(columns=[target]).select_dtypes(
    include=["object"]
).columns.tolist()

print("Numerical Columns:")
print(numerical_cols)

print("\nCategorical Columns:")
print(categorical_cols)

# Check skewness of numerical features
print("\nNumerical Feature Skewness:")
print(df_clean[numerical_cols].skew().sort_values(ascending=False))

# Check categorical distributions
print("\nCategorical Feature Distributions:")
for col in categorical_cols:
    print(f"\n{col}:")
    print(df_clean[col].value_counts())

Numerical Columns:
['Age', 'CityTier', 'DurationOfPitch', 'NumberOfPersonVisiting', 'NumberOfFollowups', 'PreferredPropertyStar', 'NumberOfTrips', 'Passport', 'PitchSatisfactionScore', 'OwnCar', 'NumberOfChildrenVisiting', 'MonthlyIncome']

Categorical Columns:
['TypeofContact', 'Occupation', 'Gender', 'ProductPitched', 'MaritalStatus', 'Designation']

Numerical Feature Skewness:
MonthlyIncome               1.746316
NumberOfTrips               1.533037
DurationOfPitch             1.384696
PreferredPropertyStar       0.901050
Passport                    0.897782
CityTier                    0.714909
Age                         0.446236
NumberOfChildrenVisiting    0.243254
NumberOfPersonVisiting     -0.022566
PitchSatisfactionScore     -0.114547
NumberOfFollowups          -0.403821
OwnCar                     -0.460543
dtype: float64

Categorical Feature Distributions:

TypeofContact:
TypeofContact
Self Enquiry       2918
Company Invited    1210
Name: count, dtype: int64

Occupation:
Occup

### Preprocessing Strategy

The dataset contains 12 numerical features and 6 categorical features after removing the identifier columns.

The categorical variables will be encoded using one-hot encoding so that they can be used by the machine learning algorithms. Unknown categories during inference will be handled safely.

The numerical features will be passed through without transformation. Although some numerical variables show positive skewness, the selected models are tree-based algorithms and generally do not require feature scaling or normally distributed numerical variables.

The target variable `ProdTaken` will be separated from the input features before model training.

In [11]:
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder

# Separate features and target
X = df_clean.drop(columns=["ProdTaken"])
y = df_clean["ProdTaken"]

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

# Identify categorical and numerical columns
categorical_cols = X.select_dtypes(include=["object"]).columns.tolist()
numerical_cols = X.select_dtypes(include=["int64", "float64"]).columns.tolist()

# Preprocessing
preprocessor = ColumnTransformer(
    transformers=[
        (
            "categorical",
            OneHotEncoder(
                handle_unknown="ignore",
                sparse_output=False
            ),
            categorical_cols
        )
    ],
    remainder="passthrough"
)

# Fit preprocessing only on training data
X_train_processed = preprocessor.fit_transform(X_train)

# Transform test data using the fitted preprocessor
X_test_processed = preprocessor.transform(X_test)

print("Training set shape:", X_train.shape)
print("Testing set shape:", X_test.shape)

print("\nProcessed training set shape:", X_train_processed.shape)
print("Processed testing set shape:", X_test_processed.shape)

print("\nTraining target distribution:")
print(y_train.value_counts(normalize=True))

print("\nTesting target distribution:")
print(y_test.value_counts(normalize=True))

Training set shape: (3302, 18)
Testing set shape: (826, 18)

Processed training set shape: (3302, 34)
Processed testing set shape: (826, 34)

Training target distribution:
ProdTaken
0    0.806784
1    0.193216
Name: proportion, dtype: float64

Testing target distribution:
ProdTaken
0    0.807506
1    0.192494
Name: proportion, dtype: float64


In [12]:
feature_names = preprocessor.get_feature_names_out()

print("Number of transformed features:", len(feature_names))

print("\nTransformed feature names:")
print(feature_names)

Number of transformed features: 34

Transformed feature names:
['categorical__TypeofContact_Company Invited'
 'categorical__TypeofContact_Self Enquiry'
 'categorical__Occupation_Free Lancer'
 'categorical__Occupation_Large Business'
 'categorical__Occupation_Salaried'
 'categorical__Occupation_Small Business' 'categorical__Gender_Female'
 'categorical__Gender_Male' 'categorical__ProductPitched_Basic'
 'categorical__ProductPitched_Deluxe' 'categorical__ProductPitched_King'
 'categorical__ProductPitched_Standard'
 'categorical__ProductPitched_Super Deluxe'
 'categorical__MaritalStatus_Divorced'
 'categorical__MaritalStatus_Married' 'categorical__MaritalStatus_Single'
 'categorical__MaritalStatus_Unmarried' 'categorical__Designation_AVP'
 'categorical__Designation_Executive' 'categorical__Designation_Manager'
 'categorical__Designation_Senior Manager' 'categorical__Designation_VP'
 'remainder__Age' 'remainder__CityTier' 'remainder__DurationOfPitch'
 'remainder__NumberOfPersonVisiting' 're

### Transformed Feature Set

The preprocessing pipeline transformed the original 18 input features into 34 model-ready features.

The six categorical variables were encoded using One-Hot Encoding, resulting in 22 binary features. The remaining 12 numerical features were retained as numerical variables.

The transformed feature set therefore contains 34 features in total. The `handle_unknown="ignore"` option was used in the encoder to ensure that unseen categorical values do not cause errors during transformation.

### Train-Test Split

The dataset was divided into 80% training data and 20% testing data, resulting in 3,302 training observations and 826 testing observations.

Stratified sampling was used to preserve the original class distribution of the target variable. The proportions of both classes remain nearly identical in the training and testing sets, which helps provide a representative evaluation of the trained models.

## Model Training and Registration with Experimentation Tracking

Multiple classification algorithms will be trained and evaluated to identify the best-performing model for predicting whether a customer will purchase the Wellness Tourism Package.

The models will be compared using Accuracy, Precision, Recall, F1-score, and ROC-AUC. ROC-AUC and F1-score will receive particular attention because the target variable is imbalanced.

In [13]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import (
    BaggingClassifier,
    RandomForestClassifier,
    AdaBoostClassifier,
    GradientBoostingClassifier
)
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)

# Define baseline models
models = {
    "Decision Tree": DecisionTreeClassifier(random_state=42),
    "Bagging": BaggingClassifier(random_state=42),
    "Random Forest": RandomForestClassifier(random_state=42),
    "AdaBoost": AdaBoostClassifier(random_state=42),
    "Gradient Boosting": GradientBoostingClassifier(random_state=42)
}

results = []

# Train and evaluate each model
for name, model in models.items():
    model.fit(X_train_processed, y_train)

    y_pred = model.predict(X_test_processed)
    y_prob = model.predict_proba(X_test_processed)[:, 1]

    results.append({
        "Model": name,
        "Accuracy": accuracy_score(y_test, y_pred),
        "Precision": precision_score(y_test, y_pred, zero_division=0),
        "Recall": recall_score(y_test, y_pred, zero_division=0),
        "F1 Score": f1_score(y_test, y_pred, zero_division=0),
        "ROC-AUC": roc_auc_score(y_test, y_prob)
    })

baseline_results = pd.DataFrame(results)

baseline_results.sort_values(
    by="ROC-AUC",
    ascending=False
).reset_index(drop=True)

,Model,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,Random Forest,0.910412,0.920792,0.584906,0.715385,0.966008
1,Bagging,0.906780,0.841667,0.635220,0.724014,0.951793
2,Gradient Boosting,0.868039,0.804878,0.415094,0.547718,0.873063
3,Decision Tree,0.888620,0.705521,0.723270,0.714286,0.825653
4,AdaBoost,0.838983,0.760000,0.238994,0.363636,0.821127


In [14]:
from sklearn.model_selection import GridSearchCV

# Hyperparameter grid for Random Forest
rf_param_grid = {
    "n_estimators": [100, 200],
    "max_depth": [None, 10, 20],
    "min_samples_split": [2, 5],
    "min_samples_leaf": [1, 2]
}

rf_grid = GridSearchCV(
    estimator=RandomForestClassifier(
        random_state=42,
        n_jobs=-1
    ),
    param_grid=rf_param_grid,
    scoring="roc_auc",
    cv=3,
    n_jobs=-1,
    verbose=1
)

rf_grid.fit(X_train_processed, y_train)

print("Best Random Forest Parameters:")
print(rf_grid.best_params_)

print("\nBest Cross-Validation ROC-AUC:")
print(rf_grid.best_score_)

Fitting 3 folds for each of 24 candidates, totalling 72 fits
Best Random Forest Parameters:
{'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 200}

Best Cross-Validation ROC-AUC:
0.935122805505479


### Random Forest Hyperparameter Tuning

Random Forest was tuned using GridSearchCV with 3-fold cross-validation and ROC-AUC as the optimization metric.

A total of 24 hyperparameter combinations were evaluated. The best configuration uses 200 estimators, no maximum depth restriction, a minimum sample split of 2, and a minimum sample leaf size of 1.

The best cross-validation ROC-AUC obtained during tuning was 0.9351.

In [15]:
# Get the best tuned Random Forest model
best_rf = rf_grid.best_estimator_

# Predictions on the test set
y_pred_rf = best_rf.predict(X_test_processed)
y_prob_rf = best_rf.predict_proba(X_test_processed)[:, 1]

# Evaluate tuned Random Forest
tuned_rf_results = {
    "Model": "Tuned Random Forest",
    "Accuracy": accuracy_score(y_test, y_pred_rf),
    "Precision": precision_score(y_test, y_pred_rf, zero_division=0),
    "Recall": recall_score(y_test, y_pred_rf, zero_division=0),
    "F1 Score": f1_score(y_test, y_pred_rf, zero_division=0),
    "ROC-AUC": roc_auc_score(y_test, y_prob_rf)
}

print("Tuned Random Forest Test Results:")
for metric, value in tuned_rf_results.items():
    if metric != "Model":
        print(f"{metric}: {value:.4f}")

Tuned Random Forest Test Results:
Accuracy: 0.9128
Precision: 0.9307
Recall: 0.5912
F1 Score: 0.7231
ROC-AUC: 0.9692


### Tuned Random Forest Evaluation

The tuned Random Forest improves upon the baseline Random Forest across all evaluated metrics.

The ROC-AUC increased from 0.9660 to 0.9692, while the F1-score increased from 0.7154 to 0.7231. Precision and recall also improved slightly.

These results indicate that the selected hyperparameters provide a better-performing Random Forest model on the unseen test data.

In [16]:
# Hyperparameter grid for Bagging
bagging_param_grid = {
    "n_estimators": [50, 100, 200],
    "max_samples": [0.5, 0.75, 1.0],
    "max_features": [0.5, 0.75, 1.0]
}

bagging_grid = GridSearchCV(
    estimator=BaggingClassifier(
        random_state=42,
        n_jobs=-1
    ),
    param_grid=bagging_param_grid,
    scoring="roc_auc",
    cv=3,
    n_jobs=-1,
    verbose=1
)

bagging_grid.fit(X_train_processed, y_train)

print("Best Bagging Parameters:")
print(bagging_grid.best_params_)

print("\nBest Cross-Validation ROC-AUC:")
print(bagging_grid.best_score_)

Fitting 3 folds for each of 27 candidates, totalling 81 fits
Best Bagging Parameters:
{'max_features': 0.75, 'max_samples': 1.0, 'n_estimators': 200}

Best Cross-Validation ROC-AUC:
0.9402107645115217


### Bagging Hyperparameter Tuning

Bagging was tuned using GridSearchCV with 3-fold cross-validation and ROC-AUC as the optimization metric.

A total of 27 hyperparameter combinations were evaluated. The best configuration uses 200 estimators, 100% of the training samples for each estimator, and 75% of the available features.

The best cross-validation ROC-AUC obtained was 0.9402.

In [17]:
# Get the best tuned Bagging model
best_bagging = bagging_grid.best_estimator_

# Predictions on the test set
y_pred_bagging = best_bagging.predict(X_test_processed)
y_prob_bagging = best_bagging.predict_proba(X_test_processed)[:, 1]

# Evaluate tuned Bagging
tuned_bagging_results = {
    "Model": "Tuned Bagging",
    "Accuracy": accuracy_score(y_test, y_pred_bagging),
    "Precision": precision_score(y_test, y_pred_bagging, zero_division=0),
    "Recall": recall_score(y_test, y_pred_bagging, zero_division=0),
    "F1 Score": f1_score(y_test, y_pred_bagging, zero_division=0),
    "ROC-AUC": roc_auc_score(y_test, y_prob_bagging)
}

print("Tuned Bagging Test Results:")
for metric, value in tuned_bagging_results.items():
    if metric != "Model":
        print(f"{metric}: {value:.4f}")

Tuned Bagging Test Results:
Accuracy: 0.9237
Precision: 0.9364
Recall: 0.6478
F1 Score: 0.7658
ROC-AUC: 0.9740


### Tuned Bagging Evaluation

The tuned Bagging model outperforms the tuned Random Forest model across all evaluation metrics.

It achieves an accuracy of 0.9237, precision of 0.9364, recall of 0.6478, F1-score of 0.7658, and ROC-AUC of 0.9740.

The improvement in recall and F1-score is particularly important because the dataset contains an imbalanced target variable. Based on the current experiments, Tuned Bagging is the strongest-performing model.

In [18]:
# Hyperparameter grid for Gradient Boosting
gb_param_grid = {
    "n_estimators": [100, 200],
    "learning_rate": [0.05, 0.1],
    "max_depth": [2, 3],
    "min_samples_split": [2, 5]
}

gb_grid = GridSearchCV(
    estimator=GradientBoostingClassifier(random_state=42),
    param_grid=gb_param_grid,
    scoring="roc_auc",
    cv=3,
    n_jobs=-1,
    verbose=1
)

gb_grid.fit(X_train_processed, y_train)

print("Best Gradient Boosting Parameters:")
print(gb_grid.best_params_)

print("\nBest Cross-Validation ROC-AUC:")
print(gb_grid.best_score_)

Fitting 3 folds for each of 16 candidates, totalling 48 fits
Best Gradient Boosting Parameters:
{'learning_rate': 0.1, 'max_depth': 3, 'min_samples_split': 5, 'n_estimators': 200}

Best Cross-Validation ROC-AUC:
0.8727289025006936


### Gradient Boosting Hyperparameter Tuning

Gradient Boosting was tuned using GridSearchCV with 3-fold cross-validation and ROC-AUC as the optimization metric.

A total of 16 hyperparameter combinations were evaluated. The best configuration uses 200 estimators, a learning rate of 0.1, maximum tree depth of 3, and a minimum sample split of 5.

The best cross-validation ROC-AUC obtained was 0.8727, which is lower than the best cross-validation scores obtained by the tuned Random Forest and Bagging models.

In [19]:
# Get the best tuned Gradient Boosting model
best_gb = gb_grid.best_estimator_

# Predictions on the test set
y_pred_gb = best_gb.predict(X_test_processed)
y_prob_gb = best_gb.predict_proba(X_test_processed)[:, 1]

# Evaluate tuned Gradient Boosting
tuned_gb_results = {
    "Model": "Tuned Gradient Boosting",
    "Accuracy": accuracy_score(y_test, y_pred_gb),
    "Precision": precision_score(y_test, y_pred_gb, zero_division=0),
    "Recall": recall_score(y_test, y_pred_gb, zero_division=0),
    "F1 Score": f1_score(y_test, y_pred_gb, zero_division=0),
    "ROC-AUC": roc_auc_score(y_test, y_prob_gb)
}

print("Tuned Gradient Boosting Test Results:")
for metric, value in tuned_gb_results.items():
    if metric != "Model":
        print(f"{metric}: {value:.4f}")

Tuned Gradient Boosting Test Results:
Accuracy: 0.8814
Precision: 0.8144
Recall: 0.4969
F1 Score: 0.6172
ROC-AUC: 0.8869


### Gradient Boosting Evaluation

The tuned Gradient Boosting model achieved an accuracy of 0.8814, precision of 0.8144, recall of 0.4969, F1-score of 0.6172, and ROC-AUC of 0.8869 on the test set.

These results are considerably lower than those of the tuned Random Forest and tuned Bagging models. Therefore, Gradient Boosting will not be selected as the final model based on the current experimentation results.

In [20]:
# Check XGBoost installation
try:
    import xgboost as xgb
    print("XGBoost version:", xgb.__version__)
except ImportError:
    print("XGBoost is not installed.")

XGBoost version: 3.2.0


### XGBoost Hyperparameter Tuning

XGBoost will also be evaluated as part of the model experimentation process. Hyperparameters will be tuned using GridSearchCV with 3-fold cross-validation, using ROC-AUC as the optimization metric.

The tuned XGBoost model will then be evaluated on the unseen test set and compared with the previously evaluated ensemble models.

In [21]:
from xgboost import XGBClassifier

# Hyperparameter grid for XGBoost
xgb_param_grid = {
    "n_estimators": [100, 200],
    "max_depth": [3, 5],
    "learning_rate": [0.05, 0.1],
    "subsample": [0.8, 1.0]
}

xgb_grid = GridSearchCV(
    estimator=XGBClassifier(
        random_state=42,
        eval_metric="logloss",
        n_jobs=-1
    ),
    param_grid=xgb_param_grid,
    scoring="roc_auc",
    cv=3,
    n_jobs=-1,
    verbose=1
)

xgb_grid.fit(X_train_processed, y_train)

print("Best XGBoost Parameters:")
print(xgb_grid.best_params_)

print("\nBest Cross-Validation ROC-AUC:")
print(xgb_grid.best_score_)

Fitting 3 folds for each of 16 candidates, totalling 48 fits
Best XGBoost Parameters:
{'learning_rate': 0.1, 'max_depth': 5, 'n_estimators': 200, 'subsample': 0.8}

Best Cross-Validation ROC-AUC:
0.9068036778473929


### XGBoost Hyperparameter Tuning

XGBoost was tuned using GridSearchCV with 3-fold cross-validation and ROC-AUC as the optimization metric.

A total of 16 hyperparameter combinations were evaluated. The best configuration uses 200 estimators, a learning rate of 0.1, a maximum tree depth of 5, and a subsample ratio of 0.8.

The best cross-validation ROC-AUC obtained was 0.9068.

In [22]:
# Get the best tuned XGBoost model
best_xgb = xgb_grid.best_estimator_

# Predictions on the test set
y_pred_xgb = best_xgb.predict(X_test_processed)
y_prob_xgb = best_xgb.predict_proba(X_test_processed)[:, 1]

# Evaluate tuned XGBoost
tuned_xgb_results = {
    "Model": "Tuned XGBoost",
    "Accuracy": accuracy_score(y_test, y_pred_xgb),
    "Precision": precision_score(y_test, y_pred_xgb, zero_division=0),
    "Recall": recall_score(y_test, y_pred_xgb, zero_division=0),
    "F1 Score": f1_score(y_test, y_pred_xgb, zero_division=0),
    "ROC-AUC": roc_auc_score(y_test, y_prob_xgb)
}

print("Tuned XGBoost Test Results:")
for metric, value in tuned_xgb_results.items():
    if metric != "Model":
        print(f"{metric}: {value:.4f}")

Tuned XGBoost Test Results:
Accuracy: 0.9262
Precision: 0.9016
Recall: 0.6918
F1 Score: 0.7829
ROC-AUC: 0.9434


### XGBoost Evaluation

The tuned XGBoost model achieved an accuracy of 0.9262, precision of 0.9016, recall of 0.6918, F1-score of 0.7829, and ROC-AUC of 0.9434 on the test set.

XGBoost achieves the highest accuracy, recall, and F1-score among the tuned models evaluated so far. However, its ROC-AUC is lower than that of the tuned Random Forest and Bagging models.

Therefore, model selection will consider multiple evaluation metrics rather than relying on a single metric.

In [23]:
# Hyperparameter grid for AdaBoost
ada_param_grid = {
    "n_estimators": [50, 100, 200],
    "learning_rate": [0.05, 0.1, 0.5, 1.0]
}

ada_grid = GridSearchCV(
    estimator=AdaBoostClassifier(random_state=42),
    param_grid=ada_param_grid,
    scoring="roc_auc",
    cv=3,
    n_jobs=-1,
    verbose=1
)

ada_grid.fit(X_train_processed, y_train)

print("Best AdaBoost Parameters:")
print(ada_grid.best_params_)

print("\nBest Cross-Validation ROC-AUC:")
print(ada_grid.best_score_)

Fitting 3 folds for each of 12 candidates, totalling 36 fits
Best AdaBoost Parameters:
{'learning_rate': 1.0, 'n_estimators': 200}

Best Cross-Validation ROC-AUC:
0.8131784579889483


### AdaBoost Hyperparameter Tuning

AdaBoost was tuned using GridSearchCV with 3-fold cross-validation and ROC-AUC as the optimization metric.

A total of 12 hyperparameter combinations were evaluated. The best configuration uses 200 estimators and a learning rate of 1.0.

The best cross-validation ROC-AUC obtained was 0.8132, which is substantially lower than the tuned Random Forest, Bagging, and XGBoost models.

In [24]:
# Get the best tuned AdaBoost model
best_ada = ada_grid.best_estimator_

# Predictions on the test set
y_pred_ada = best_ada.predict(X_test_processed)
y_prob_ada = best_ada.predict_proba(X_test_processed)[:, 1]

# Evaluate tuned AdaBoost
tuned_ada_results = {
    "Model": "Tuned AdaBoost",
    "Accuracy": accuracy_score(y_test, y_pred_ada),
    "Precision": precision_score(y_test, y_pred_ada, zero_division=0),
    "Recall": recall_score(y_test, y_pred_ada, zero_division=0),
    "F1 Score": f1_score(y_test, y_pred_ada, zero_division=0),
    "ROC-AUC": roc_auc_score(y_test, y_prob_ada)
}

print("Tuned AdaBoost Test Results:")
for metric, value in tuned_ada_results.items():
    if metric != "Model":
        print(f"{metric}: {value:.4f}")

Tuned AdaBoost Test Results:
Accuracy: 0.8426
Precision: 0.7164
Recall: 0.3019
F1 Score: 0.4248
ROC-AUC: 0.8261


### AdaBoost Evaluation

The tuned AdaBoost model achieved an accuracy of 0.8426, precision of 0.7164, recall of 0.3019, F1-score of 0.4248, and ROC-AUC of 0.8261 on the test set.

The model performs considerably worse than the tuned Random Forest, Bagging, Gradient Boosting, and XGBoost models. In particular, its low recall and F1-score indicate that it is not effective at identifying customers who are likely to purchase the package.

Therefore, AdaBoost will not be selected as the final model.

In [25]:
# Compile results from all tuned models
experiment_results = pd.DataFrame([
    tuned_rf_results,
    tuned_bagging_results,
    tuned_gb_results,
    tuned_xgb_results,
    tuned_ada_results
])

# Display models ranked by F1-score
experiment_results = experiment_results.sort_values(
    by="F1 Score",
    ascending=False
).reset_index(drop=True)

experiment_results

,Model,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,Tuned XGBoost,0.926150,0.901639,0.691824,0.782918,0.943377
1,Tuned Bagging,0.923729,0.936364,0.647799,0.765799,0.973985
2,Tuned Random Forest,0.912833,0.930693,0.591195,0.723077,0.969166
3,Tuned Gradient Boosting,0.881356,0.814433,0.496855,0.617188,0.886906
4,Tuned AdaBoost,0.842615,0.716418,0.301887,0.424779,0.826078


### Final Model Selection

Based on the experimentation results, XGBoost is selected as the final model.

XGBoost achieved the highest accuracy (0.9262), recall (0.6918), and F1-score (0.7829) among the tuned models. These metrics are particularly relevant for the business problem because the objective is to identify customers who are likely to purchase the Wellness Tourism Package before contacting them.

Although the tuned Bagging model achieved a higher ROC-AUC (0.9740) and precision (0.9364), XGBoost provides a better balance between identifying potential purchasers and maintaining prediction quality, as reflected by its higher recall and F1-score.

Therefore, the tuned XGBoost model will be used as the final model for registration and deployment.

In [26]:
# Display the final model configuration
print("Final Model: XGBoost")
print("\nBest Parameters:")
print(xgb_grid.best_params_)

print("\nFinal Test Metrics:")
print(f"Accuracy : {tuned_xgb_results['Accuracy']:.4f}")
print(f"Precision: {tuned_xgb_results['Precision']:.4f}")
print(f"Recall   : {tuned_xgb_results['Recall']:.4f}")
print(f"F1 Score : {tuned_xgb_results['F1 Score']:.4f}")
print(f"ROC-AUC  : {tuned_xgb_results['ROC-AUC']:.4f}")

Final Model: XGBoost

Best Parameters:
{'learning_rate': 0.1, 'max_depth': 5, 'n_estimators': 200, 'subsample': 0.8}

Final Test Metrics:
Accuracy : 0.9262
Precision: 0.9016
Recall   : 0.6918
F1 Score : 0.7829
ROC-AUC  : 0.9434


### Experiment Tracking

MLflow is used to track the machine learning experiments. For each tuned model, the selected hyperparameters and evaluation metrics are recorded.

This provides a reproducible record of the experiments and allows the final model to be identified based on the recorded performance.

In [27]:
# Check MLflow installation
try:
    import mlflow
    print("MLflow version:", mlflow.__version__)
except ImportError:
    print("MLflow is not installed.")

MLflow version: 3.16.0


In [28]:
import importlib.metadata as md
import pydantic
import mlflow

print("typing_extensions:", md.version("typing_extensions"))
print("pydantic:", pydantic.__version__)
print("MLflow:", mlflow.__version__)

typing_extensions: 4.16.0
pydantic: 2.13.5
MLflow: 3.16.0


In [29]:
import mlflow
from pathlib import Path

# Create a local SQLite database for MLflow tracking
mlflow_db = Path.cwd() / "mlflow.db"

mlflow.set_tracking_uri(f"sqlite:///{mlflow_db.as_posix()}")

# Create/select the experiment
mlflow.set_experiment("Visit_with_Us_Tourism_Prediction")

print("MLflow tracking URI:", mlflow.get_tracking_uri())
print("MLflow experiment tracking initialized successfully.")

MLflow tracking URI: sqlite:///C:/Users/sanjeev kumar/Desktop/python project/tourism_project/data/mlflow.db
MLflow experiment tracking initialized successfully.


### Logging Tuned XGBoost Experiment

The tuned XGBoost model is logged in MLflow along with its hyperparameters and evaluation metrics. This experiment represents the selected final model based on the comparison of tuned algorithms.

In [30]:
import mlflow

with mlflow.start_run(run_name="Tuned_XGBoost"):
    mlflow.log_params({
        "n_estimators": 200,
        "max_depth": 5,
        "learning_rate": 0.1,
        "subsample": 0.8
    })

    mlflow.log_metrics({
        "accuracy": 0.926150,
        "precision": 0.901639,
        "recall": 0.691824,
        "f1_score": 0.782918,
        "roc_auc": 0.943377
    })

    print("Tuned XGBoost experiment logged successfully.")
    print("Run ID:", mlflow.active_run().info.run_id)

Tuned XGBoost experiment logged successfully.
Run ID: 8f7f60405d01441692308150ca9e79ca


In [31]:
import mlflow

experiments = {
    "Tuned_Bagging": {
        "params": {
            "n_estimators": 200,
            "max_samples": 1.0,
            "max_features": 0.75
        },
        "metrics": {
            "accuracy": 0.923729,
            "precision": 0.936364,
            "recall": 0.647799,
            "f1_score": 0.765799,
            "roc_auc": 0.973985
        }
    },
    "Tuned_Random_Forest": {
        "params": {
            "n_estimators": 200,
            "max_depth": "None",
            "min_samples_split": 2,
            "min_samples_leaf": 1
        },
        "metrics": {
            "accuracy": 0.912833,
            "precision": 0.930693,
            "recall": 0.591195,
            "f1_score": 0.723077,
            "roc_auc": 0.969166
        }
    },
    "Tuned_Gradient_Boosting": {
        "params": {
            "n_estimators": 200,
            "learning_rate": 0.1,
            "max_depth": 3,
            "min_samples_split": 5
        },
        "metrics": {
            "accuracy": 0.881356,
            "precision": 0.814433,
            "recall": 0.496855,
            "f1_score": 0.617188,
            "roc_auc": 0.886906
        }
    },
    "Tuned_AdaBoost": {
        "params": {
            "n_estimators": 200,
            "learning_rate": 1.0
        },
        "metrics": {
            "accuracy": 0.842615,
            "precision": 0.716418,
            "recall": 0.301887,
            "f1_score": 0.424779,
            "roc_auc": 0.826078
        }
    }
}

for model_name, experiment in experiments.items():
    with mlflow.start_run(run_name=model_name):
        mlflow.log_params(experiment["params"])
        mlflow.log_metrics(experiment["metrics"])

        print(f"{model_name} logged successfully.")
        print("Run ID:", mlflow.active_run().info.run_id)

Tuned_Bagging logged successfully.
Run ID: ab57d7a5c8304c09ba04d94adbbcf95d
Tuned_Random_Forest logged successfully.
Run ID: db3360dbe9e44693b399b546c4d6dafc
Tuned_Gradient_Boosting logged successfully.
Run ID: 60080a08fdc94bcab65c9a7258ea0ad3
Tuned_AdaBoost logged successfully.
Run ID: ccd83c42d95f441e940cd98b95bb7ff2


In [32]:
import mlflow
import pandas as pd

experiment = mlflow.get_experiment_by_name(
    "Visit_with_Us_Tourism_Prediction"
)

runs = mlflow.search_runs(
    experiment_ids=[experiment.experiment_id],
    order_by=["metrics.f1_score DESC"]
)

runs[[
    "run_id",
    "tags.mlflow.runName",
    "metrics.accuracy",
    "metrics.precision",
    "metrics.recall",
    "metrics.f1_score",
    "metrics.roc_auc"
]]

,run_id,tags.mlflow.runName,metrics.accuracy,metrics.precision,metrics.recall,metrics.f1_score,metrics.roc_auc
0,8f7f60405d01441692308150ca9e79ca,Tuned_XGBoost,0.926150,0.901639,0.691824,0.782918,0.943377
1,6a22474c224241459867e8c6d2b6f210,Tuned_XGBoost,0.926150,0.901639,0.691824,0.782918,0.943377
2,f6c60d221445425cbcd56f96e4c82c89,Tuned_XGBoost,0.926150,0.901639,0.691824,0.782918,0.943377
3,a64b812aabfb4a19a3daa04b57a5e50c,Tuned_XGBoost,0.926150,0.901639,0.691824,0.782918,0.943377
4,ab57d7a5c8304c09ba04d94adbbcf95d,Tuned_Bagging,0.923729,0.936364,0.647799,0.765799,0.973985
5,fc0ffbbf222644f490b8fa79a208fa87,Tuned_Bagging,0.923729,0.936364,0.647799,0.765799,0.973985
6,b56407e5610f465988a5cbe68ecb7e72,Tuned_Bagging,0.923729,0.936364,0.647799,0.765799,0.973985
7,c99b5058117148b29a39203218af11d1,Tuned_Bagging,0.923729,0.936364,0.647799,0.765799,0.973985
8,db3360dbe9e44693b399b546c4d6dafc,Tuned_Random_Forest,0.912833,0.930693,0.591195,0.723077,0.969166
9,0afc8a9bf75a40a89f7103c8f75c3960,Tuned_Random_Forest,0.912833,0.930693,0.591195,0.723077,0.969166


### Experiment Tracking

MLflow is used to track the machine learning experiments. For each tuned model, the selected hyperparameters and evaluation metrics are recorded.

This provides a reproducible record of the experiments and allows the final model to be identified based on the recorded performance.

# Deployment

In [33]:
import joblib
from sklearn.pipeline import Pipeline

# Create the final preprocessing + model pipeline
final_pipeline = Pipeline([
    ("preprocessor", preprocessor),
    ("model", xgb_grid.best_estimator_)
])

# Fit the complete pipeline on the training data
final_pipeline.fit(X_train, y_train)

print("Final pipeline trained successfully.")

Final pipeline trained successfully.


In [34]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score

y_pred = final_pipeline.predict(X_test)
y_prob = final_pipeline.predict_proba(X_test)[:, 1]

print("Pipeline Test Metrics")
print("---------------------")
print(f"Accuracy : {accuracy_score(y_test, y_pred):.4f}")
print(f"Precision: {precision_score(y_test, y_pred):.4f}")
print(f"Recall   : {recall_score(y_test, y_pred):.4f}")
print(f"F1 Score : {f1_score(y_test, y_pred):.4f}")
print(f"ROC-AUC  : {roc_auc_score(y_test, y_prob):.4f}")

Pipeline Test Metrics
---------------------
Accuracy : 0.9262
Precision: 0.9016
Recall   : 0.6918
F1 Score : 0.7829
ROC-AUC  : 0.9434


In [35]:
import os

os.makedirs("tourism_project/deployment", exist_ok=True)

print("Deployment folder created successfully.")

Deployment folder created successfully.


In [36]:
import joblib

model_path = "../deployment/final_model.joblib"

joblib.dump(final_pipeline, model_path)

print(f"Model saved successfully at: {model_path}")

Model saved successfully at: ../deployment/final_model.joblib


## Dockerfile

In [37]:
os.makedirs("tourism_project/deployment", exist_ok=True)

In [38]:
%%writefile ../deployment/Dockerfile
# Use a minimal base image with Python 3.9 installed
FROM python:3.9

# Set the working directory inside the container to /app
WORKDIR /app

# Copy all files from the current directory on the host to the container's /app directory
COPY . .

# Install Python dependencies listed in requirements.txt
RUN pip3 install -r requirements.txt

RUN useradd -m -u 1000 user
USER user
ENV HOME=/home/user \
    PATH=/home/user/.local/bin:$PATH

WORKDIR $HOME/app

COPY --chown=user . $HOME/app

# Define the command to run the Streamlit app
CMD ["streamlit", "run", "app.py", "--server.port=8501", "--server.address=0.0.0.0", "--server.enableXsrfProtection=false"]

Overwriting ../deployment/Dockerfile


## Streamlit App

Please ensure that the web app script is named `app.py`.

In [39]:
%%writefile ../deployment/app.py

import streamlit as st
import pandas as pd
import joblib
from huggingface_hub import hf_hub_download

# -----------------------------
# Load model from Hugging Face
# -----------------------------
MODEL_REPO = "rhythm0511/tourism-purchase-prediction"

model_path = hf_hub_download(
    repo_id=MODEL_REPO,
    filename="final_model.joblib",
    repo_type="model"
)

model = joblib.load(model_path)

# -----------------------------
# Streamlit UI
# -----------------------------
st.set_page_config(
    page_title="Wellness Tourism Package Prediction",
    page_icon="🏖️"
)

st.title("🏖️ Wellness Tourism Package Prediction")
st.write("Predict whether a customer is likely to purchase the package.")

st.subheader("Customer Information")

age = st.number_input("Age", min_value=18, max_value=100, value=30)
type_of_contact = st.selectbox(
    "Type of Contact",
    ["Self Enquiry", "Company Invited"]
)
city_tier = st.selectbox("City Tier", [1, 2, 3])
occupation = st.selectbox(
    "Occupation",
    ["Salaried", "Free Lancer", "Small Business", "Large Business"]
)
gender = st.selectbox("Gender", ["Male", "Female"])

number_of_person_visiting = st.number_input(
    "Number of Persons Visiting", min_value=1, value=2
)

preferred_property_star = st.selectbox(
    "Preferred Property Star", [3, 4, 5]
)

marital_status = st.selectbox(
    "Marital Status",
    ["Single", "Divorced", "Married", "Unmarried"]
)

number_of_trips = st.number_input(
    "Number of Trips", min_value=0, value=2
)

passport = st.selectbox("Passport", [0, 1])
own_car = st.selectbox("Own Car", [0, 1])

number_of_children_visiting = st.number_input(
    "Number of Children Visiting", min_value=0, value=0
)

designation = st.selectbox(
    "Designation",
    ["Manager", "Executive", "Senior Manager", "AVP", "VP"]
)

monthly_income = st.number_input(
    "Monthly Income", min_value=0.0, value=20000.0
)

pitch_satisfaction_score = st.selectbox(
    "Pitch Satisfaction Score", [1, 2, 3, 4, 5]
)

product_pitched = st.selectbox(
    "Product Pitched",
    ["Deluxe", "Basic", "Standard", "Super Deluxe", "King"]
)

number_of_followups = st.number_input(
    "Number of Followups", min_value=0, value=3
)

duration_of_pitch = st.number_input(
    "Duration of Pitch", min_value=0, value=15
)

if st.button("Predict Purchase"):
    
    input_data = pd.DataFrame([{
        "Age": age,
        "TypeofContact": type_of_contact,
        "CityTier": city_tier,
        "Occupation": occupation,
        "Gender": gender,
        "NumberOfPersonVisiting": number_of_person_visiting,
        "PreferredPropertyStar": preferred_property_star,
        "MaritalStatus": marital_status,
        "NumberOfTrips": number_of_trips,
        "Passport": passport,
        "OwnCar": own_car,
        "NumberOfChildrenVisiting": number_of_children_visiting,
        "Designation": designation,
        "MonthlyIncome": monthly_income,
        "PitchSatisfactionScore": pitch_satisfaction_score,
        "ProductPitched": product_pitched,
        "NumberOfFollowups": number_of_followups,
        "DurationOfPitch": duration_of_pitch
    }])

    prediction = model.predict(input_data)[0]
    probability = model.predict_proba(input_data)[0][1]

    if prediction == 1:
        st.success(
            f"Customer is likely to purchase the package "
            f"(Probability: {probability:.2%})"
        )
    else:
        st.warning(
            f"Customer is unlikely to purchase the package "
            f"(Probability: {probability:.2%})"
        )

Overwriting ../deployment/app.py


## Dependency Handling

Please ensure that the dependency handling file is named `requirements.txt`.

In [40]:
%%writefile ../deployment/requirements.txt
streamlit
pandas
numpy
scikit-learn
xgboost
joblib
huggingface_hub

Overwriting ../deployment/requirements.txt


In [41]:
import subprocess
import sys

subprocess.Popen([
    sys.executable,
    "-m",
    "streamlit",
    "run",
    "../deployment/app.py",
    "--server.port=8501"
])

print("Streamlit app started on port 8501.")

Streamlit app started on port 8501.


# Hosting

In [42]:
%pip install -q huggingface_hub

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 23.0.1 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [43]:
from huggingface_hub import notebook_login

notebook_login()

In [44]:
from huggingface_hub import HfApi

api = HfApi()

HF_USERNAME = "rhythm0511"
MODEL_REPO = "tourism-purchase-prediction"

repo_id = f"{HF_USERNAME}/{MODEL_REPO}"

# Create Hugging Face Model repository
api.create_repo(
    repo_id=repo_id,
    repo_type="model",
    exist_ok=True
)

# Upload the trained model
api.upload_file(
    path_or_fileobj="../deployment/final_model.joblib",
    path_in_repo="final_model.joblib",
    repo_id=repo_id,
    repo_type="model"
)

print("Model uploaded successfully!")
print(f"Model repository: https://huggingface.co/{repo_id}")

Processing Files (0 / 0): |          |  0.00B /  0.00B            

New Data Upload: |          |  0.00B /  0.00B            

No files have been modified since last commit. Skipping to prevent empty commit.


Model uploaded successfully!
Model repository: https://huggingface.co/rhythm0511/tourism-purchase-prediction


# MLOps Pipeline with Github Actions Workflow

**Note:**

1. Before running the file below, make sure to add the HF_TOKEN to your GitHub secrets to enable authentication between GitHub and Hugging Face.
2. The below code is for a sample YAML file that can be updated as required to meet the requirements of this project.

```
name: Tourism Project Pipeline

on:
  push:
    branches:
      - main  # Automatically triggers on push to the main branch

jobs:

  register-dataset:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: Install Dependencies
        run: <add_code_here>
      - name: Upload Dataset to Hugging Face Hub
        env:
          HF_TOKEN: ${{ secrets.HF_TOKEN }}
        run: <add_code_here>

  data-prep:
    needs: register-dataset
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: Install Dependencies
        run: <add_code_here>
      - name: Run Data Preparation
        env:
          HF_TOKEN: ${{ secrets.HF_TOKEN }}
        run: <add_code_here>


  model-traning:
    needs: data-prep
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: Install Dependencies
        run: <add_code_here>
      - name: Start MLflow Server
        run: |
          nohup mlflow ui --host 0.0.0.0 --port 5000 &  # Run MLflow UI in the background
          sleep 5  # Wait for a moment to let the server starts
      - name: Model Building
        env:
          HF_TOKEN: ${{ secrets.HF_TOKEN }}
        run: <add_code_here>


  deploy-hosting:
    runs-on: ubuntu-latest
    needs: [model-traning,data-prep,register-dataset]
    steps:
      - uses: actions/checkout@v3
      - name: Install Dependencies
        run: <add_code_here>
      - name: Push files to Frontend Hugging Face Space
        env:
          HF_TOKEN: ${{ secrets.HF_TOKEN }}
        run: <add_code_here>

```

**Note:** To use this YAML file for our use case, we need to

1. Go to the GitHub repository for the project
2. Create a folder named ***.github/workflows/***
3. In the above folder, create a file named ***pipeline.yml***
4. Copy and paste the above content for the YAML file into the ***pipeline.yml*** file

## Requirements file for the Github Actions Workflow

In [45]:
!git --version

git version 2.55.0.windows.5


In [47]:
import os

os.chdir("..")
print(os.getcwd())

C:\Users\sanjeev kumar\Desktop\python project\tourism_project


In [48]:
!git init

Initialized empty Git repository in C:/Users/sanjeev kumar/Desktop/python project/tourism_project/.git/


In [50]:
!git config --global user.name "rhythm-05"
!git config --global user.email "rhythmgaba511@gmail.com"

In [51]:
!git config --global user.name
!git config --global user.email

rhythm-05
rhythmgaba511@gmail.com


## Github Authentication and Push Files

In [52]:
!git remote add origin https://github.com/rhythm-05/tourism-purchase-prediction-mlops.git

In [53]:
!git remote -v

origin	https://github.com/rhythm-05/tourism-purchase-prediction-mlops.git (fetch)
origin	https://github.com/rhythm-05/tourism-purchase-prediction-mlops.git (push)


In [54]:
!git status --short

?? .ipynb_checkpoints/
?? Learner_Template_Notebook_AML_and_MLOps_Project.ipynb
?? data/
?? deployment/


In [55]:
%%writefile .gitignore

.ipynb_checkpoints/
__pycache__/
*.pyc
*.db
.env
.venv/
venv/

Writing .gitignore


In [56]:
!git status --short

?? .gitignore
?? Learner_Template_Notebook_AML_and_MLOps_Project.ipynb
?? data/
?? deployment/


In [57]:
import os

print("DATA FOLDER:")
print(os.listdir("data"))

print("\nDEPLOYMENT FOLDER:")
print(os.listdir("deployment"))

DATA FOLDER:
['.ipynb_checkpoints', 'Learner_Template_Notebook_AML_and_MLOps_Project.ipynb', 'mlflow.db', 'mlruns', 'tourism.csv', 'tourism_project']

DEPLOYMENT FOLDER:
['app.py', 'Dockerfile', 'final_model.joblib', 'requirements.txt']


In [58]:
%%writefile .gitignore

.ipynb_checkpoints/
__pycache__/
*.pyc
*.db
mlruns/
.env
.venv/
venv/

Overwriting .gitignore


In [65]:
!git add .

In [66]:
!git status

On branch master

No commits yet

Changes to be committed:
  (use "git rm --cached <file>..." to unstage)
	new file:   .gitignore
	new file:   Learner_Template_Notebook_AML_and_MLOps_Project.ipynb
	new file:   data/Learner_Template_Notebook_AML_and_MLOps_Project.ipynb
	new file:   data/tourism.csv
	new file:   deployment/Dockerfile
	new file:   deployment/app.py
	new file:   deployment/final_model.joblib
	new file:   deployment/requirements.txt



In [67]:
!git rm --cached "data/Learner_Template_Notebook_AML_and_MLOps_Project.ipynb"

rm 'data/Learner_Template_Notebook_AML_and_MLOps_Project.ipynb'


In [70]:
!git status

On branch master

No commits yet

Changes to be committed:
  (use "git rm --cached <file>..." to unstage)
	new file:   .gitignore
	new file:   Learner_Template_Notebook_AML_and_MLOps_Project.ipynb
	new file:   data/tourism.csv
	new file:   deployment/Dockerfile
	new file:   deployment/app.py
	new file:   deployment/final_model.joblib
	new file:   deployment/requirements.txt



In [71]:
!git commit -m "Initial MLOps project setup"

[master (root-commit) f403a3f] Initial MLOps project setup
 7 files changed, 7466 insertions(+)
 create mode 100644 .gitignore
 create mode 100644 Learner_Template_Notebook_AML_and_MLOps_Project.ipynb
 create mode 100644 data/tourism.csv
 create mode 100644 deployment/Dockerfile
 create mode 100644 deployment/app.py
 create mode 100644 deployment/final_model.joblib
 create mode 100644 deployment/requirements.txt


In [72]:
!git branch -M main

In [73]:
!git branch

* main


In [77]:
!git status

On branch main
nothing to commit, working tree clean


* Before moving forward, we need to generate a secret token to push files directly from Colab to the GitHub repository.
* Please follow the below instructions to create the GitHub token:
    - Open your GitHub profile.
    - Click on ***Settings***.
    - Go to ***Developer Settings***.
    - Expand the ***Personal access tokens*** section and select ***Tokens (classic)***.
    - Click ***Generate new token***, then choose ***Generate new token (classic)***.
    - Add a note and select all required scopes.
    - Click ***Generate token***.
    - Copy the generated token and store it safely in a notepad.

In [ ]:
# Install Git
!apt-get install git

# Set your Git identity (replace with your details)
!git config --global user.email "<-------GitHub Email Address------->"
!git config --global user.name "<--------GitHub UserName--------->"

# Clone your GitHub repository
!git clone https://github.com/<--------GitHub UserName--------->/<--------GitHub Reponame--------->.git

# Move your folder to the repository directory
!mv /content/tourism_project/ /content/<--------GitHub Reponame--------->

In [ ]:
# Change directory to the cloned repository
%cd <--------GitHub Reponame--------->/

# Add the new folder to Git
!git add .

# Commit the changes
!git commit -m "first commit"

# Push to GitHub (you'll need your GitHub credentials; use a personal access token if 2FA enabled)
!git push https://<--------GitHub UserName--------->:<--------GitHub Token--------->@github.com/<--------GitHub UserName--------->/<--------GitHub Reponame--------->.git

# Output Evaluation

- GitHub (link to repository, screenshot of folder structure and executed workflow)

- Streamlit on Hugging Face (link to HF space, screenshot of Streamlit app)

<font size=6 color="navyblue">Power Ahead!</font>
___